### DISCUSSION TITLE
Tinyuf2: Support using TinyUF2 with MicroPython.

### DISCUSSION THREAD

**Discussion by robert-hh** [adafruit/tinyuf2#413] [Q&A]
TinyUF2 is a good project and can make firmware loading pretty easy. So I started to join MicroPython Tiny UF2 with some progress. What I did on MicroPython after some analysis of the TinyUF2 code:
- I moved the Interrupt vector table of 0x6000C000 for the MIMXRT10xx devces, and the .text segment up to 0x6000C400.
- I added code to the startup code at  ResetHandler to copy the Interrupt vector table from 0c6000C000 to the start of the ITCM segment at address 0.
- I added code to the software bootloader activation to write a magic number to the SNVS->LPGPR[3] register and then Reset the device. That allows as well to activate the bootloader with the touch1200 method.
- I created .uf2 files of the MicroPython firmware which skip all segments below 0x6000C000.

That makes TinyUF2 kind of working for MIMXRT1011, MIMXRT1015, MIXRT1020 and MIMXRT1052 board, as far as the UF2 loader starts, the MicroPython firmware is accepted and will be executed. There are a few minor restrictions:  
With the MIMXRT1015EVK and Teensy, the software bootloader does not work. Something happens, but the bootloader does not expose the boot drive. And getting the bootloader to pop up with double reset requires a fine rhythm. But that is more related to the IMXRT10xx not having a clean reset pin. 

My primary question is about another topic which should be solved first:

With all devices the flash is write locked when the UF2 bootloader is used. So is there a mechanism in the bootloader that locks the flash, which has to be unlocked?

**Comment by robert-hh** [comment]
The problem with the write lock is solved by changing code in MicroPython, using the same flash configuration as TinyUF2.

The problem which still exists is with the MIMXRT1015EVK board and Teensy (MIMXRT1062) when starting the bootloader by software: write a magic number to the SNVS->LPGPR[3] register and then Reset the device. That works for the bootloader versions of  MIMXRT1011, -1052 and -1052 but at MIXRT1015, -1062 (Teensy). The board resets, but the bootloader does create the bootloader drive. It seems to start, but it's unclear in which state it is. When I hook up a JLing debugger, it crashes. Message:
`Reason: A fault with configurable priority has been escalated to a HardFault exception at 0x20008000.`
That address is beyond the MIMXRT1015 DTCM space, which is configured for 32K.
The PC is at 2020_9948, which looks like a data table, and is according inside of /usr/lib/gcc/arm-none-eabi/12.2.1/thumb/v7e-m+dp/hard/crtbegin.o.
SP is at 2000_7f90.

**Reply by robert-hh** [reply]
Problem kind of solved for MIMXRT1015 by using GCC version 10.3 instead of 12.1. Problem persists for Teensy4x. In that zombie state lsusb shows `NXP Semiconductors SE Blank RT Family`.

**Reply by robert-hh** [reply]
The problem with Teensy exists with Circuitpython, too. So I'd better do not spend more time on it.

**Reply by hathach** [reply]
hi @robert-hh sorry for late response, I am glad to hear you are incorporating tinyuf2 to micropython. Changing gcc seems to suggest there is optimization/aligment and/or timing issue. I am currently working on other works. Give me a bit of time, I will pull out my 1015 and try to figure out what is wrong. For Teensy, I have no idea, but tinyuf2 work well with rt1062 evk, so I guess there is something in the hardware differences.

**Reply by robert-hh** [reply]
hi @hathach . First of all, thank you for your great work on tinyUSB and tinyUF2. It is a big support for the community. And do not worry. Sometimes I just have to write down my concerns to be readable by other people to sort my thoughts. With the 1015 and the compiler version: I had made a similar experience with the UF2 bootloader for SAMD, where the authors state that it only works when compiled with GCC 10.3. Which I could confirm. That's why I tested GCC 10.3 with the 1015 version.
That change did not work for the Teensy. It seems simply not to properly reset when done by software with e.g. NVIC_SystemReset(). There may be conflict with the built-in Teensy Loader support. At this state, lsusb shows `NXP Semiconductors SE Blank RT Family`. The the chip seems to be in the ROM bootloader mode. But I do not  mind. The Teensy Loader is present and works.
Thank you for confirming, that it works with the MIMXRT1060-EVK board. I do not own it and could not test. Same for MIMXRT1064-EVK and MIMXRT1024-EVK.

**Reply by hathach** [reply]
thank @robert-hh for your kind words, I am really possitive that rt1064/60 should work. It is used as one of main imxrt dev board by me and @tannewt (for circuitpython). For 1015, I will pull it out and test next time I switch to tinyuf2 project.

**Comment by water5** [comment]
I install MicroPython on ESP32-S3 use [https://micropython.org/resources/firmware/ESP32_GENERIC_S3-20250415-v1.25.0.uf2](https://micropython.org/resources/firmware/ESP32_GENERIC_S3-20250415-v1.25.0.uf2) via TinyUF2, but can't into UF2 bootloader mode (the BOOT drive not appear) after `machine.bootloader()`.

@https://github.com/orgs/micropython/discussions/17525
